lib/Controller/AvatarController.php

```diff
@@ -0,0 +1,129 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @copyright Copyright (c) 2022, Vitor Mattos <vitor@php.rio>
+ *
+ * @author Vitor Mattos <vitor@php.rio>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Talk\Controller;
+
+use InvalidArgumentException;
+use OC\Files\Filesystem;
+use OCA\Talk\Service\AvatarService;
+use OCP\AppFramework\Http;
+use OCP\AppFramework\Http\DataResponse;
+use OCP\AppFramework\Http\FileDisplayResponse;
+use OCP\AppFramework\Http\Response;
+use OCP\Files\IAppData;
+use OCP\IL10N;
+use OCP\IRequest;
+use OCP\IUserSession;
+
+class AvatarController extends AEnvironmentAwareController {
+	private AvatarService $avatarService;
+	private IUserSession $userSession;
+	private IL10N $l;
+
+	public function __construct(
+		string $appName,
+		IRequest $request,
+		IAppData $appData,
+		AvatarService $avatarService,
+		IUserSession $userSession,
+		IL10N $l10n
+	) {
+		parent::__construct($appName, $request);
+		$this->appData = $appData;
+		$this->avatarService = $avatarService;
+		$this->userSession = $userSession;
+		$this->l = $l10n;
+	}
+
+	/**
+	 * @PublicPage
+	 * @RequireModeratorParticipant
+	 */
+	public function uploadAvatar(): DataResponse {
+		try {
+			$file = $this->request->getUploadedFile('file');
+
+			if (is_null($file)) {
+				throw new InvalidArgumentException($this->l->t('No image file provided'));
+			}
+
+			if (
+				$file['error'] !== 0 ||
+				!is_uploaded_file($file['tmp_name']) ||
+				Filesystem::isFileBlacklisted($file['tmp_name'])
+			) {
+				throw new InvalidArgumentException($this->l->t('Invalid file provided'));
+			}
+			if ($file['size'] > 20 * 1024 * 1024) {
+				throw new InvalidArgumentException($this->l->t('File is too big'));
+			}
+
+			$content = file_get_contents($file['tmp_name']);
+			unlink($file['tmp_name']);
```

Comment:
should move all this to the service?
I kind of dislike reading the image content and passing it to a method

Reply:
The first part of this code is too much specific of handling file from global var `$_FILE` and I think that is best to maintain in the controller.

A good point is about send the content of image or the path of image to `setImage` method

I think that the current way is the best way sending the content of image to method because we need do delete the uploaded file. The method `service->setImage` throw exceptions and this can broke the flow and don't delete the file.

Do you have any suggestion?

Reply:
You could either create the OC_Image here and forward that on the message call, or you create a `setAvatarFromRequest` which then does all the meta handling and then internally calls setAvatar with "okay" parameters?